src/toolbox/xor.rs

```diff
@@ -0,0 +1,270 @@
+        match xor_many_scalars(secret, &shares) {
+            Ok(val) => shares.push(val),
+            Err(e) => return Err(e),
+        }
+
+        return Ok(shares);
+    }
+
+    /// For Xor, sparse_shares should include a None value for each share you want us to generate
+    fn complete(&mut self, secret: &Scalar, sparse_shares: &Vec<Option<Scalar>>) -> Result<Vec<Scalar>, String> {
+        let mut empties = 0;
+        let mut shares = Vec::new();
+        let mut new_shares = Vec::new();
+        for share in sparse_shares.iter() {
+            if share.is_some() {
+                shares.push(share.unwrap());
+            } else {
+                if empties != 0 {     // the "zero-th" empty will be the one we have to XOR-calculate
+                    trace!("Pushing a random Scalar to extra empty");
+                    new_shares.push(Scalar::random(&mut self.rng));
+                }
+                empties += 1;
+            }
+        }
+
+        if empties < 1 {
+            return Err(String::from("No empty shares in which to complete!"));
+        }
+
+        info!("Completing {} empties with {} known and secret = {:?}", empties, shares.len(), secret);
+
+        match xor_many_more_scalars(secret, &shares, &new_shares) {
+            Ok(val) => {
+                debug!("The XOR'd share is {:?}", val);
+                new_shares.push(val);
+            },
+            Err(e) => return Err(e),
+        }
+
+        let ret_shares: Vec<Scalar> = sparse_shares.iter().map(|x| {
+            if x.is_some() {
+                x.unwrap()
+            } else {
+                // it's OK to panic if we get a None here, since that should *never* happen
+                new_shares.pop().unwrap()
+            }
+        }).collect();
+
+        return Ok(ret_shares);
+    }
+
+    fn reconstruct(&mut self, shares: &Vec<Scalar>) -> Result<Scalar, String> {
+        if shares.len() < 1 {
+            return Err(String::from("No shares provided, impossible to reconstruct!"));
+        }
+
+        let secret = xor_many_scalars(&shares[0], &shares[1..]);
+        
+        debug!("Used {} shares to reconstruct {:?}", shares.len(), secret);
+        return secret;
+    }
+}
+
+/// Given two Scalar values, return the XOR of their constituent byte arrays
+pub fn xor_scalars(a: &Scalar, b: &Scalar) -> Scalar {
+    let mut the_bytes = [0 as u8; 32];
+    let abytes = a.to_bytes();
+    let bbytes = b.to_bytes();
+    for i in 0..the_bytes.len() {
+        the_bytes[i] = abytes[i] ^ bbytes[i];
+    }
+    Scalar::from_bits(the_bytes)
+}
+
+/// Given a list of scalars, XOR them all together.  We split out the `first` separately for efficiency's sake;
+/// there are use cases where the initial value you want to XOR with isn't in your Vector, so you don't need to
+/// push it.
+/// 
+/// TODO WARNING: the result isn't guaranteed to be modulo group order.  Is that going to be a problem?
```

Comment:
I suppose the `Scalar::from_bits` will actual assure that the resulting element is within the group? Or won't it parse a bit-string that is not resulting in a correct scalar? I would suggest investigating this with some tests on `Scalar::from_bits` giving it some large and small test byte arrays.

Reply:
The documentation is explicit that `from_bits` does not make that guarantee.  Other constructors, such as `from_canonical_bytes`, will take that extra step, but the function used here assumes you know exactly what you're doing with each bit in the data structure.

I believe that allowing the library to reduce things modulo group order would break things, but to be honest, I don't know enough of the math involved.  It just seems "risky" to me to mess with bit patterns during XOR operations.  I could write tests on that, if you like?

Reply:
Yes, I think some tests will help here. So, what does it do when you do a from_bits on something that is not within the group, it errs or do you get a Scalar outside the group? The latter would be strange though. Anyway, we need to end up in the group, so making sure we actually take the element modulo the order of the group is the right way to go. It would be deterministic as well, which is what we need.

Reply:
`from_bits()` gives you an object of type `Scalar` which may or may not be in the group.  You can test it with `is_canonical()`.

I think what we want is `complete()` doing no checking, but have `reconstruct()` error out if the returned Scalar isn't in the group.  The XOR method itself doesn't need to worry about group-ness.  Thoughts?